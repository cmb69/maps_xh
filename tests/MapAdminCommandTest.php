<?php

namespace Maps;

use ApprovalTests\Approvals;
use Maps\Model\Map;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class MapAdminCommandTest extends TestCase
{
    private DocumentStore $store;
    private View $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $map = Map::update("london", $this->store);
        $map->addMarker(0, 0, "basic info", true);
        $this->store->commit();
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["maps"]);
    }

    private function sut(): MapAdminCommand
    {
        return new MapAdminCommand($this->store, $this->view);
    }

    public function testRendersOverview(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()($request);
        $this->assertSame("Maps – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersEditorForNewMap(): void
    {
        $request = new FakeRequest(["url" => "http://example.com/?&maps&admin=plugin_main&action=create"]);
        $response = $this->sut()($request);
        $this->assertSame("Maps – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testCreatesNewMap(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=create",
            "post" => [
                "maps_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertFileExists(vfsStream::url("root/new.xml"));
        $this->assertSame("http://example.com/?&maps&admin=plugin_main", $response->location());
    }

    public function testReportsFailureToSaveNewMap(): void
    {
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=create",
            "post" => [
                "maps_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the map!", $response->output());
    }

    public function testRendersEditorForUpdate(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
        ]);
        $response = $this->sut()($request);
        $this->assertSame("Maps – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testUpdatesMap(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
            "post" => [
                "maps_do" => "",
                "markers" => "0|0|more info|",
            ],
        ]);
        $response = $this->sut()($request);
        $map = Map::retrieve("london", $this->store);
        $this->assertCount(1, $map->markers());
        $this->assertSame("http://example.com/?&maps&admin=plugin_main&maps_map=london", $response->location());
    }

    public function testReportsFailureToUpdateMap(): void
    {
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
            "post" => [
                "maps_do" => "",
                "markers" => "0|0|more info|",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the map!", $response->output());
    }
}
