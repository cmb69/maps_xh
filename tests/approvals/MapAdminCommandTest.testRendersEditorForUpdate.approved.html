
<article class="map_edit">
  <h1>Maps â€“ Administration</h1>
  <form method="post">
    <label>
      <span>Name</span>
      <span class="maps_help">(only a-z, 0-9, and hyhpens)</span>
      <input name="name" value="london" disabled required pattern="[a-z0-9\-]+">
    </label>
    <label>
      <span>Latitude</span>
      <span class="maps_help">(between -90 and 90)</span>
      <input type="number" name="latitude" value="0" min="-90" max="90" step="any">
    </label>
    <label>
      <span>Longitude</span>
      <span class="maps_help">(between -180 and 180)</span>
      <input type="number" name="longitude" value="0" min="-180" max="180" step="any">
    <label>
      <span>Zoom</span>
      <span class="maps_help">(between 0 and 20)</span>
      <input type="number" name="zoom" value="0" min="0" max="20">
    </label>
    <label>
      <span>Max. Zoom</span>
      <span class="maps_help">(between 0 and 20)</span>
      <input type="number" name="max_zoom" value="0" min="0" max="20">
    </label>
    <label>
      <span>Markers</span>
      <span class="maps_help">(latitude|longitude|info|show; one marker per line)</span>
      <textarea name="markers">0|0|basic info|1
</textarea>
    </label>
    <p class="maps_controls">
      <button name="maps_do">Save</button>
    </p>
  </form>
</article>
